judge/views/problem.py

```diff
@@ -201,6 +201,93 @@ def get_context_data(self, **kwargs):
                                           context['description'], 'problem')
         context['meta_description'] = self.object.summary or metadata[0]
         context['og_image'] = self.object.og_image or metadata[1]
+
+        context['can_vote'] = self.object.can_vote(user)
+        if context['can_vote']:
+            try:
+                context['vote'] = ProblemPointsVote.objects.get(voter=user.profile, problem=self.object)
+            except ObjectDoesNotExist:
+                context['vote'] = None
+        else:
+            context['vote'] = None
+
+        return context
+
+
+class ProblemVote(ProblemMixin, DetailView):
+    context_object_name = 'problem'
+    template_name = 'problem/vote-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if not self.object.can_vote(self.request.user):
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        try:
+            context['vote'] = ProblemPointsVote.objects.get(voter=self.request.profile, problem=self.object)
+        except ObjectDoesNotExist:
+            context['vote'] = None
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = max(settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE,
+                                           settings.DMOJ_PROBLEM_MIN_PROBLEM_POINTS)
+        return context
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if not problem.can_vote(request.user):
+            return HttpResponseForbidden(_('Not allowed to vote on this problem.'), content_type='text/plain')
+
+        form = ProblemPointsVoteForm(request.POST)
+        if not form.is_valid():
+            return JsonResponse(form.errors, status=400)
+
+        with transaction.atomic():
+            # Delete any pre-existing votes.
+            ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+            vote = form.save(commit=False)
+            vote.voter = request.profile
+            vote.problem = problem
+            vote.save()
+        return JsonResponse({'points': vote.points})
+
+
+class DeleteProblemVote(ProblemMixin, SingleObjectMixin, View):
+    http_method_names = ['options', 'post']  # This disables GET requests, even though ProblemMixin.get exists.
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if not problem.can_vote(request.user):
+            return HttpResponseForbidden(_('Not allowed to delete votes on this problem.'), content_type='text/plain')
+
+        ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+        return HttpResponse('success', content_type='text/plain')
+
+
+class ProblemVoteStats(ProblemMixin, DetailView):
+    context_object_name = 'problem'
+    template_name = 'problem/vote-stats-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if self.request.in_contest:
```

Comment:
Is this the only check we need here? Should we move this check to a model object?

Reply:
i created `Problem.get_vote_perm` to check both the view perm and edit perm
i refactored all the permission checks, please review them again